src/nvim/api/ui_events.in.h

```diff
@@ -123,6 +123,9 @@ void win_viewport(Integer grid, Window win, Integer topline,
                   Integer line_count)
   FUNC_API_SINCE(7) FUNC_API_REMOTE_ONLY;
 
+void win_extmarks(Window win, Array data)
```

Comment:
This doesn't need to be an array, but could actually spell the named arguments `win, ns_id, mark_id, start_row, end_row, start_col`. The UI msgpack layer already packs events of the same kind very efficiently as long as their are sent together (at one place in `win_update()`, after all lines, in this case). Only downside is we need a separate `win_extmarks_clear` event but I think it worth able to use event metadata as intended.

Reply:
why `win_extmarks_clear` and what's it for?

Reply:
I mean an empty event in case there were extmarks before but now they are now invalid. but perhaps that won't be necessary, actually.  the UI knows after a win_line that old extmarks on that line is no longer valid, regardless if new ones are supplied or not.

Reply:
I just implemented this in FVim -- an extmark is attached to a grid cell, and then if the cell is invalidated (e.g. scrolled out of view) with no further win_extmarks updates, the ui finds the inconsistency and hide the marks.
There's one bad case though: in insert mode, if a new line is input and it triggers a scroll (e.g. enter at the bottom row), and the top row has the same content as the second top row (e.g. two empty lines), there's no grid event for the top rows 😅
Guess that's when `win_extmarks_clear` should come in, but I'm not sure where to start implementing this...

Reply:
I think I get what you mean now -- just like how grid_line combines events from multiple grids.

Reply:
done.